Less string replacement

tellowkrinkle opened the pull request:
- Logging no longer uses `%@` + varargs.  Swift string interpolation should be used instead.  This prevents accidents where you include args but forget to include a `%@` in the string.  The calls are now lazy as well, so any string conversion you do will only be run if the result is going to be printed.
- `JSON.encodeJSONData` no longer converts the data to a string and back to data.  It seemed kind of unnecessary.
- Endpoint generation now uses enums with associated types and a separate method which generates urls instead of string replacement with a dictionary.  This means that you won't be able to forget a required piece or mistype its name.
- `DiscordEnpoint.createRequest` now takes the HTTP method and any data associated with it and fills the request object accordingly, so it should be much harder to accidentally leave one of those out (and much less to do in any method that makes requests).
- New `executeRequest` method for `DiscordRateLimiter` which takes an endpoint (with its associated data), a token, an HTTP method (and its associated data), and a callback and does everything else for you.
- New rate limit key creation which (I think) conforms to the spec listed on Discord's website.

Note: If you want to add a new endpoint, there are now three places where you have to add code.  They're all default-less switch statements, so the compiler should complain at you if you forget one.  One is the enum definition, one is in the enum's description method (where it actually creates the URL), and the third is in DiscordRateLimitKey's init (to indicate what parts to use as the rate limit key).

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -27,113 +70,171 @@ import Foundation
+
+    public var description: String {
+        switch self {
+        case .baseURL: return "https://discordapp.com/api/v6"
```
Just put these on a new line

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -143,33 +244,66 @@ public enum DiscordEndpoint : String {
+        if case let .get(params) = method {
+            getParams = params
+        }
+        else {
```
Else should be on the same line as the }

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -143,33 +244,66 @@ public enum DiscordEndpoint : String {
+        if case let .post(optionalContent) = method {
+            content = optionalContent
+        }
+        else if case let .put(optionalContent) = method {
```
Else should be on the same line as }

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -143,33 +244,66 @@ public enum DiscordEndpoint : String {
+        if let content = content {
+            request.httpBody = content.0
+            request.setValue(content.type.description, forHTTPHeaderField: "Content-Type")
+            request.setValue(content.0.count.description, forHTTPHeaderField: "Content-Length")
```
Use the string constructor, not description.

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -17,6 +17,49 @@
+/**
+ * An HTTP Request Method.  This includes any associated data.
+ */
+public enum HTTPMethod {
```
Something named HTTPMethod shouldn't have the actual data in it as well, it's basically a request at that point.

TellowKrinkle replied:
Do you have any thoughts on the name?  I don't want to call it a request because its data gets added to an existing request object, and running `createRequest` and giving it a request would be confusing.

nuclearace replied:
Maybe move this inside of DiscordEndpoint and rename it EndpointRequest? Maybe it would make sense to move `createRequest` to be a method of this and pass in the endpoint?

TellowKrinkle replied:
All right

nuclearace commented on Sources/SwiftDiscord/DiscordRateLimiter.swift:
```diff
@@ -18,8 +18,8 @@
 
-private typealias RequestCallback = (Data?, HTTPURLResponse?, Error?) -> ()
-private typealias RateLimitedRequest = (request: URLRequest, callback: RequestCallback)
+public typealias DiscordRequestCallback = (Data?, HTTPURLResponse?, Error?) -> ()
```
Does this need to be public or can it be left private or internal?

TellowKrinkle commented:
My thought on the rate limit key change is that we already had to specify a different endpoint for the rate limit key from the request endpoint when making requests, so for each new endpoint, there's still work to be done (even if it gets deferred to when you actually use the endpoint), and I'd rather you do it when you're setting up the endpoint, since then you'll only have to do it once (and you can't make a mistake later and give the wrong endpoint for the rate limit key).  Previously I had just added a method to DiscordEndpoint to get the appropriate rate limit key, but then I realized that some rate limit keys were more unique than (what I understood of) the spec, including things like invite codes and stuff, which wastes memory (since all of these keys are put in as keys to a dictionary) even if it doesn't make a difference to Discord (since invites probably aren't rate limited).

Also, since I am planning on making Snowflake IDs into UInt64s instead of strings (which uses about 1/10 the memory per unique Snowflake in my tests), that will make a rate limit key use 16 bytes of memory compared to the 16 bytes + UTF16 string (I assume the internal storage of String is Length + Pointer to storage of string) that it would take up if it were stored as a string.

nuclearace commented:
I'm still not convinced on the rate limiter change. Can that be spliced out for another time?

TellowKrinkle commented:
Sure

nuclearace commented:
Also the UInt64 for snowflakes is a good idea.

I'm hoping Swift 4 and its Codeable protocol makes all the custom deserialization I have to do easier.

TellowKrinkle commented:
In that case, how should we separate the message send endpoint from the message delete endpoint?  (From the discord API documentation: `There is currently a single exception to the above rule regarding different HTTP methods sharing the same rate limit, and that is for the deletion of messages. Deleting messages falls under a separate, higher rate limit so that bots are able to more quickly delete content from channels (which is useful for moderation bots).`)

nuclearace commented:
Yeah I knew about that exception, I just didn't care enough to add a case for it. Still don't really care.

TellowKrinkle commented:
All right.  I added a separate case for it anyways (I just put a check in the initializer to handle it separately)

nuclearace commented:
Also fun fact; I lifted the logging system used here from [my other library](https://github.com/socketio/socket.io-client-swift). Very early on it actually used string interpolation. But I moved to this system because I didn't want to do the interpolation if the string wasn't going to be printed (and I didn't know about `@autoclosure` since this was basically when Swift came out).

nuclearace commented on Sources/SwiftDiscord/DiscordRateLimiter.swift:
```diff
@@ -145,14 +156,13 @@ public struct DiscordRateLimitKey: Hashable {
+		if case let .channelMessageDelete(channel, _) = endpoint {
+			self.key = DiscordEndpoint.messages(channel: channel).description + "d"
+		}
+		else {
```
Missed one

nuclearace commented on Sources/SwiftDiscord/DiscordEndpoint.swift:
```diff
@@ -27,149 +44,381 @@ import Foundation
+			if case let .get(params) = self {
+				getParams = params
+			}
+			else {
```
Sorry, but I'm a bit nitpicky when it comes to style, esp with things like brackets.

TellowKrinkle commented:
All right, I fixed those.  I thought I had gotten all those, but apparently I introduced new ones after I checked.  I also fixed my accidental use of tabs instead of spaces.
